query_ranges_to_vnodes.cc

```diff
@@ -108,6 +108,13 @@ void query_ranges_to_vnodes_generator::process_one_range(size_t n, dht::partitio
     }
 
     if (ranges.size() < n) {
-        add_range(get_remainder());
+        auto range = get_remainder();
+        const bool is_empty = range.start().has_value() && range.end().has_value() &&
+            range.start()->value().relation_to_keys() - range.end()->value().relation_to_keys() == 2 &&
+            dht::token::to_int64(range.end()->value().token()) - dht::token::to_int64(range.start()->value().token()) == 1 &&
```

Comment:
Two keys can have the same token.

Reply:
Here I only check for the case described in the commit message:
* the user provided a range in the form `[dht::ring_position::ending_at(endpoint_token), dht::ring_position::starting_at(next_token))`;
* after the main loop and split at `endpoint_token` we end up with a range `(dht::ring_position::starting_at(next_token), dht::ring_position::starting_at(next_token))` as a remainder

This last range's `ring_position`s have `token_bound`, not keys, this's checked by the condition `range.start()->value().relation_to_keys() - range.end()->value().relation_to_keys() == 2`.

Reply:
We need a comment explaining why this check is enough.
If `(dht::ring_position::starting_at(next_token), dht::ring_position::starting_at(next_token))` is the only case we are interested in, can the check be simpler?

Reply:
Actually, `dht::ring_position::ending_at` as a start and `dht::ring_position::starting_at` as an end position seems invalid and it would be better to place an assert somewhere that this is not the case.

Reply:
That is a wrapping range and we do have an assert for that.

Reply:
Wrapping range is when an end boundary is greater than the start boundary, but I mean `ring_position` with `token_bound::end` used as a start boundary. From user perspective this start boundary can be equivalently expressed as `{token+1, token_bound::start}`.

Reply:
It may or may not be a wrapping range, depending on the inclusiveness. Being able to express inclusiveness on two levels was a big mistake, it makes everything so confusing and opens up questions like this that we shouldn't even be able to ask.

Reply:
> Being able to express inclusiveness on two levels was a big mistake, it makes everything so confusing and opens up questions like this that we shouldn't even be able to ask.

I have the same feeling.

I updated the PR (and the commit message) and made the check more direct. The idea is that in this particular code we don't care about empty ranges in the input, it's the client responsibility to check for this, but we strive not to add additional empty ranges as an artifact of the algorithm.